### PR TITLE
Add Imagenav baseline into main + update object search in semantic map

### PR DESCRIPTION
**Pull request by cpaxton**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: projects/stretch_objectnav/eval_episode.py [hunk 1]
```diff
@@ -18,7 +17,6 @@
     rospy.init_node("eval_episode_stretch_objectnav")
     agent = ObjectNavAgent(config=config)
     env = StretchObjectNavEnv(config=config)
-    env.goto(STRETCH_HOME_Q)
```
Comment (cpaxton) [inline projects/stretch_objectnav/eval_episode.py:21]: Why these changes in stretch obj nav?

File: .gitmodules [hunk 2]
```diff
@@ -1,15 +1,15 @@
 [submodule "hab_stretch"]
 	path = assets/hab_stretch
 	url = git@github.com:cpaxton/hab_stretch.git
-[submodule "third_party/contact_graspnet"]
```
[the remark this answers is not in the document]
Comment (theophilegervet) [reply .gitmodules:4]: Did I remove this? My bad, we should add it back

File: projects/stretch_objectnav/eval_episode.py (hunk 1 above)
Comment (theophilegervet) [reply projects/stretch_objectnav/eval_episode.py:21]: This is now in the `reset()` method of the Stretch environment

Comment (cpaxton) [reply projects/stretch_objectnav/eval_episode.py:21]: sg

File: .gitmodules (hunk 2 above)
[the remark this answers is not in the document]
Comment (cpaxton) [reply .gitmodules:4]: yeah do you think you could revert this? then i think i can sign off on the PR

File: src/home_robot/home_robot/agent/imagenav_agent/visualizer.py [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+        visited_mask = self.visited_map_vis[gy1:gy2, gx1:gx2] == 1
+        semantic_map[no_category_mask] = 0
+        semantic_map[np.logical_and(no_category_mask, explored_mask)] = 2
+        semantic_map[np.logical_and(no_category_mask, obstacle_mask)] = 1
+        semantic_map[visited_mask] = 3
+
+        # Goal
+        if visualize_goal:
+            selem = skimage.morphology.disk(4)
+            goal_mat = 1 - skimage.morphology.binary_dilation(goal_map, selem) != True
```
Comment (cpaxton) [inline src/home_robot/home_robot/agent/imagenav_agent/visualizer.py:488]: @theophilegervet i have no idea what this is doing but the linter does not like it

File: src/home_robot/home_robot/agent/imagenav_agent/visualizer.py [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+            goal_mask = goal_mat == 1
+            semantic_map[goal_mask] = 21
+            if closest_goal_map is not None:
+                closest_goal_mat = (
+                    1 - skimage.morphology.binary_dilation(closest_goal_map, selem)
+                    != True
```
Comment (cpaxton) [inline src/home_robot/home_robot/agent/imagenav_agent/visualizer.py:494]: @theophilegervet same here - i have no idea what this is doing but the linter does not like it

File: .gitignore [hunk 5]
```diff
@@ -30,3 +30,5 @@ datadump/
 *.h5
 *.hdf5
 *.mp4
+
+src/home_robot/home_robot/agent/imagenav_agent/SuperGluePretrainedNetwork
```
Comment (exhaustin) [inline .gitignore:34]: No need to add this here if it's now included as a submodule.

File: projects/instanceimagenav/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,4 @@
+scikit-fmm==2022.8.15
+scikit-learn==0.22.2.post1
+scikit-image==0.15.0
+trimesh
```
Comment (exhaustin) [inline projects/instanceimagenav/requirements.txt:4]: I had to manually install `natsort` when running the code. Should it be also listed here?

File: projects/instanceimagenav/README.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+Components:
+
+- Habitat v0.2.3 (Sim & Lab)
+- Detic
+- SuperGlue
+- HM3D Scene Dataset
+- 2023 Habitat Instance ImageNav Challenge Dataset
+
+```bash
+git clone -b imagenav --recurse-submodules git@github.com:jacobkrantz/home-robot.git
```
Comment (exhaustin) [inline projects/instanceimagenav/README.md:20]: Update instructions to pull from facebookresearch instead.

Comment (exhaustin) [inline projects/instanceimagenav/README.md:20]: Imo we should also remind users to `git submodule update --init --recursive` if they have `home_robot` installed already.

File: src/home_robot/home_robot/core/interfaces.py [hunk 8]
```diff
@@ -55,6 +55,7 @@ class Observations:
     semantic: Optional[
         np.array
     ] = None  # (camera_height, camera_width) in [0, num_sem_categories - 1]
+    camera_extrinsic: Optional[np.array] = None  # (4, 4) camera pose in world coordinates
```
Comment (exhaustin) [inline src/home_robot/home_robot/core/interfaces.py:58]: Maybe just `camera_pose`? So it would be coherent with other fields that might be added to the Observation (`base_pose`, `ee_pose`, etc.)

File: src/home_robot/home_robot/utils/rotation.py [hunk 9]
```diff
@@ -92,7 +92,7 @@ def get_grid(pose, grid_size, precision):
     theta22 = torch.stack([torch.zeros_like(x), torch.ones_like(x), y], 1)
     theta2 = torch.stack([theta21, theta22], 1)
 
-    rot_grid = F.affine_grid(theta1, torch.Size(grid_size)).to(precision)
-    trans_grid = F.affine_grid(theta2, torch.Size(grid_size)).to(precision)
+    rot_grid = F.affine_grid(theta1, torch.Size(grid_size), align_corners=False).to(precision)
+    trans_grid = F.affine_grid(theta2, torch.Size(grid_size), align_corners=False).to(precision)
```
Comment (exhaustin) [inline src/home_robot/home_robot/utils/rotation.py:96]: Would be a good time to add type hints to this function

File: src/home_robot/home_robot/perception/detection/detic/detic_mask.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+        self.augment_mask_with_box = config.augment_mask_with_box
+        self.visualize = visualize
+        num_classes = len(self.metadata.thing_classes)
+        self.cpu_device = torch.device("cpu")
+
+        self.predictor = DefaultPredictor(self.setup_cfg(config))
+        reset_cls_test(self.predictor.model, classifier, num_classes)
+
+    @staticmethod
+    def _get_clip_embeddings(vocabulary, prompt="a "):
```
Comment (exhaustin) [inline src/home_robot/home_robot/perception/detection/detic/detic_mask.py:61]: Add type hints (to this and multiple other functions in this file)

File: src/home_robot_hw/home_robot_hw/env/stretch_image_nav_env.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+            self.forward_step = config.habitat.simulator.forward_step_size  # in meters
+            self.rotate_step = np.radians(config.habitat.simulator.turn_angle)
+            self.image_goal = self._load_image_goal(config.stretch_goal_image_path)
+        else:
+            self.forward_step = 0.25
+            self.rotate_step = np.radians(30)
+            self.image_goal = None
+        self.reset()
+
+    def _load_image_goal(self, goal_img_path) -> np.ndarray:
```
Comment (exhaustin) [inline src/home_robot_hw/home_robot_hw/env/stretch_image_nav_env.py:31]: Add type hints (to this and multiple other functions in this file)

File: src/home_robot_sim/home_robot_sim/env/habitat_imagenav_env/habitat_imagenav_env.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+from home_robot_sim.env.habitat_abstract_env import HabitatEnv
+
+
+MIN_DEPTH_REPLACEMENT_VALUE = 10000
+MAX_DEPTH_REPLACEMENT_VALUE = 10001
+
+
+class HabitatImageNavEnv(HabitatEnv):
+
+    def __init__(self, habitat_env: habitat.core.env.Env, config):
```
Comment (exhaustin) [inline src/home_robot_sim/home_robot_sim/env/habitat_imagenav_env/habitat_imagenav_env.py:17]: Add type hint for `config`

**Review by exhaustin — COMMENTED** [review]
One thing I would like to discuss: `objectnav_agent` and `imagenav_agent` each contain an agent with numerous modules, most of which are very specific to the particular setup in their respective projects. Maybe we should move the agents out of `home_robot` and into `projects/`? Only components that are potentially reusable by other projects (such as the frontier exploration policy) stays within `home_robot`.
Any thoughts?

File: src/home_robot_hw/home_robot_hw/env/stretch_image_nav_env.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+        goal_image = goal_image[:, :, ::-1]
+        assert goal_image.shape[0] == 512
+        assert goal_image.shape[1] == 512
+        return goal_image
+
+    def reset(self):
+        self._episode_start_pose = xyt2sophus(self.get_base_pose())
+        self.goto(STRETCH_HOME_Q)
+
+    def apply_action(self, action: Action):
```
Comment (zephirefaith) [inline src/home_robot_hw/home_robot_hw/env/stretch_image_nav_env.py:43]: docstring explaining what action means, how it should be formatted would be great.

File: src/home_robot_sim/home_robot_sim/env/habitat_imagenav_env/habitat_imagenav_env.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+import home_robot
+from home_robot_sim.env.habitat_abstract_env import HabitatEnv
+
+
+MIN_DEPTH_REPLACEMENT_VALUE = 10000
+MAX_DEPTH_REPLACEMENT_VALUE = 10001
+
+
+class HabitatImageNavEnv(HabitatEnv):
```
Comment (zephirefaith) [inline src/home_robot_sim/home_robot_sim/env/habitat_imagenav_env/habitat_imagenav_env.py:15]: general comment for this file: docstrings + type hints for code going in main `src/` repositories.

**Comment by jacobkrantz** [comment]
Rebased
Addressed comments (thanks for the reviews!)
Ensured simulation performance is as expected
Reused existing home-robot modules (eg `DiscretePlanner`)

File: .gitmodules (hunk 2 above)
[the remark this answers is not in the document]
Comment (cpaxton) [reply .gitmodules:4]: @jacobkrantz was this fixed?

[the remark this answers is not in the document]
Comment (jacobkrantz) [reply .gitmodules:4]: `contact_graspnet` was listed twice, so this was a de-dup

**Review by exhaustin — COMMENTED** [review]
Looks good in general
1. Need to rebase main again.
2. Fix linting tests (can be fixed by simply running pre-commit).
3. Fix other tests which are breaking, not sure what the problem is, let me know if I should investigate.
4. Would love to get some thoughts on moving the project-specific agents and envs out of `src/home_robot` and into `projects/`, as mentioned in my above comment?

**Comment by jacobkrantz** [comment]
@exhaustin 
1. done
2. fixed
3. I don't understand the CI errors and I can't reproduce them locally. Could you take a look?
4. discussion continued on slack

**Review by exhaustin — APPROVED** [review]
> 3. I don't understand the CI errors and I can't reproduce them locally. Could you take a look?

Seems to be broken on other open PRs as well. Smells like some upstream dependency update broke this.
I'll look into this, don't worry about it for this PR, but please ensure that `pytest tests/` pass locally for you.
